### DISCUSSION TITLE
How to convert vertex coordinates correctly in shader?

### DISCUSSION THREAD

**Discussion by yhl533556** [stride3d/stride#1440] [Q&A] [answered]
i just want do step-by-step with this tutorial:  https://roystan.net/articles/grass-shader.html
![屏幕截图 2022-06-14 123956](https://user-images.githubusercontent.com/19642756/173494872-a7a8a350-30fe-4db0-92e2-ef83f6d35502.png)
but ,the position is not right. i donot know why?
shader Test17XuKongShengCao01 :ShaderBase,ComputeColor , Texturing ,Transformation
{
     stream bool colorGreen;
     stream float4 vertex : POSITION;
     stream float3 normal : NORMAL;  
     stream float4 tangent :TANGENT;

float4 vertexOutput(float4 pos0, float3 posX)
{
    float4 pos=mul(float4(posX,1),WorldViewProjection);
    return pos+pos0;
}

    [maxvertexcount(21)]
    void GSMain(triangle  Input input[3] , inout TriangleStream<Output> triStream)
    {
                float3 vNormal =input[0].normal;
                float4 vTangent =input[0].tangent;
                float3 vBinormal=cross(vNormal,vTangent)*vTangent.w;

                float3x3 tangentToLcal= float3x3(
                	vTangent.x, vBinormal.x, vNormal.x,
                	vTangent.y, vBinormal.y, vNormal.y,
                	vTangent.z, vBinormal.z, vNormal.z
                	);

                streams=input[0];                
                float4 posss0=streams.ShadingPosition;              
                streams.colorGreen=false;
                triStream.Append(streams);

                streams=input[1];
                streams.colorGreen=false;
                triStream.Append(streams);

                streams=input[2];
                streams.colorGreen=false;
                triStream.Append(streams);

                float4 f1=float4(-0.1, 0, 0 ,1);
                float4 f2=float4( 0.1, 0,0 ,1);
                float4 f3=float4(0, 0,1 ,1);

 
                float3 posss1=mul(tangentToLcal,f1.xyz);
                float3 posss2=mul(tangentToLcal,f2.xyz);
                float3 posss3=mul(tangentToLcal,f3.xyz);

                streams.ShadingPosition=vertexOutput(posss0,posss1);
                streams.colorGreen=true;
                triStream.Append(streams);
      
                streams.ShadingPosition=vertexOutput(posss0,posss2);
                streams.colorGreen=true;
                triStream.Append(streams);
                       
                streams.ShadingPosition=vertexOutput(posss0,posss3);
                streams.colorGreen=true;
                triStream.Append(streams);
    }

        override stage void PSMain()
    {
            if(streams.colorGreen)
            {
                streams.ColorTarget=float4(1,50,1,1);
            }
            else
            {
                streams.ColorTarget = float4(1,1,1,1);
            }
    }
};

        public override void Update()
        {
            this.Entity.Get<ModelComponent>().Materials[0].Passes[0].Parameters.Set<Matrix>(WorldViewProjection, this.Entity.Get<TransformComponent>().WorldMatrix * Camera.ViewProjectionMatrix);
        }

**Comment by tebjan** [comment, marked as answer]
The `ShadingPosition` is usually already in screen space if you have a regular vertex shader.

`float4 posss0=streams.ShadingPosition;`

so maybe try:

`float4 posss0=streams.Position;`

because you multiply by `WorldViewPostion` in the `vertexOutput` method.
